### ISSUE TITLE
Bug: macro parsing fail

### ISSUE THREAD

**Issue by tcew** [libocca/occa#170]
I was having trouble getting an okl kernel to compile, and suspected the associated device function was problematic. So I switched back to a macro. 

This:

```cpp
#define surfaceTerms(sk, m, i,j , sgeo, x, y, z, vmapM, mapB, s_q, s_ndq) \
    const dlong  idM = vmapM[sk];                       \
const dfloat nx  = sgeo[sk*p_Nsgeo+p_NXID];     \
const dfloat ny  = sgeo[sk*p_Nsgeo+p_NYID];     \
const dfloat nz  = sgeo[sk*p_Nsgeo+p_NZID];             \
const dfloat WsJ = sgeo[sk*p_Nsgeo+p_WSJID];                    \
dfloat dudxP=0, dudyP=0, dudzP=0, uP=0; \
const int bc = mapB[idM];                                               \
s_q  [m][j][i]  = uP;                                                   \
s_ndq[m][j][i]  = -WsJ*(nx*dudxP + ny*dudyP + nz*dudzP);                 \

@kernel void ellipticRhsBCHex3D(const dlong Nelements,
                                @restrict const  dfloat *  ggeo,
                                @restrict const  dfloat *  sgeo,
                                @restrict const  dfloat *  D,
...
```

Compiles to the kernel launcher file as:

```cpp
...
#include <cstdlib>
#include <stdint.h>
using namespace std;
using namespace occa;
**const double nx = sgeo[sk * 14 + 0];
const double ny = sgeo[sk * 14 + 1];
const double nz = sgeo[sk * 14 + 7];
const double WsJ = sgeo[sk * 14 + 5];
double dudxP = 0, dudyP = 0, dudzP = 0, uP = 0;
const int bc = mapB[idM];
s_q[m][j][i] = uP;
s_ndq[m][j][i] = -WsJ * (nx * dudxP + ny * dudyP + nz * dudzP);**

extern "C" void ellipticRhsBCHex3D(occa::kernel_v **deviceKernel,
                                   const int &Nelements,
                                   const double *ggeo,
                                   const double *sgeo,
                                   const double *D,
                                   const double *S,
```

Overall, I suspect that the parser is failing for both macros and for device functions.

**Comment by dmed256** [comment]
Can you link the full kernel, I can't reproduce it from just that

```cpp
#define dlong int
#define dfloat float

#define surfaceTerms(sk, m, i,j , sgeo, x, y, z, vmapM, mapB, s_q, s_ndq) \
    const dlong  idM = vmapM[sk];                       \
const dfloat nx  = sgeo[sk*p_Nsgeo+p_NXID];     \
const dfloat ny  = sgeo[sk*p_Nsgeo+p_NYID];     \
const dfloat nz  = sgeo[sk*p_Nsgeo+p_NZID];             \
const dfloat WsJ = sgeo[sk*p_Nsgeo+p_WSJID];                    \
dfloat dudxP=0, dudyP=0, dudzP=0, uP=0; \
const int bc = mapB[idM];                                               \
s_q  [m][j][i]  = uP;                                                   \
s_ndq[m][j][i]  = -WsJ*(nx*dudxP + ny*dudyP + nz*dudzP);                 \

@kernel void ellipticRhsBCHex3D(const dlong Nelements,
                                @restrict const  dfloat *  ggeo,
                                @restrict const  dfloat *  sgeo,
                                @restrict const  dfloat *  D) {
  for (int i = 0; i < 10; ++i; @tile(16, @outer, @inner)) {
  }
}
```
&darr;
```cpp
> occa translate --mode Serial test.okl
#include <occa.hpp>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <stdint.h>
using namespace std;
using namespace occa;

extern "C" void ellipticRhsBCHex3D(const int &Nelements,
                                   const float *ggeo,
                                   const float *sgeo,
                                   const float *D) {
  for (int _occa_tiled_i = 0; _occa_tiled_i < 10; _occa_tiled_i += 16) {
    for (int i = _occa_tiled_i; i < (_occa_tiled_i + 16); ++i) {
      if (i < 10) {}
    }
  }
}
```

**Comment by tcew** [comment]
Almost all problems with holmes & occa 1.0 appear to be resolved - albeit using Noel's fork of occa 1.0.

One remaining issue is the lack of a portable atomic add for integers. This impacts isosurface extraction where the number of triangles in the extracted surface is not known a priori and atomic incrementing of a triangle counter is particularly useful.
